### PR TITLE
feat: add relay close logic on long inactivity

### PR DESCRIPTION
**Pull request by ice-ajax**
## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Refactoring
- [ ] Documentation
- [ ] Chore

### CODE REVIEW

**Review by ice-kreios — APPROVED** [review]

File: lib/app/features/nostr/providers/relays_provider.c.dart [hunk 1]
```diff
@@ -9,11 +11,55 @@ part 'relays_provider.c.g.dart';
...
 
 @Riverpod(keepAlive: true)
-Future<NostrRelay> relay(Ref ref, String url) async {
-  // TODO: think how to implement the following
-  //
-  // Close a connection if ALL of the following happens:
-  // 1. Has no active subscriptions
-  // 2. it has been at least 30s since last message from the relay was received
-  return NostrRelay.connect(url);
+class RelayNotifier extends _$RelayNotifier {
```
[the remark this answers is not in the document]
Comment (ice-ajax) [reply lib/app/features/nostr/providers/relays_provider.c.dart:14]: done

File: lib/app/utils/restartable_timer.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: ice License 1.0
+
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+
+class RestartableTimer implements Timer {
```
[the remark this answers is not in the document]
Comment (ice-ajax) [reply lib/app/utils/restartable_timer.dart:7]: done, moved to services

File: lib/app/features/nostr/providers/relays_provider.c.dart [hunk 3]
```diff
@@ -9,11 +11,55 @@ part 'relays_provider.c.g.dart';
...
+  late RestartableTimer _timer;
+  late StreamSubscription<RelayMessage> _messagesSubscription;
+  late StreamSubscription<int> _subscriptionsCountSubscription;
+  bool _timerFinished = false;
```
[the remark this answers is not in the document]
Comment (ice-ajax) [reply lib/app/features/nostr/providers/relays_provider.c.dart:18]: you're right, refactored

File: lib/app/features/nostr/providers/relays_provider.c.dart [hunk 4]
```diff
@@ -9,11 +11,55 @@ part 'relays_provider.c.g.dart';
...
+  @override
+  Future<NostrRelay> build(String url) async {
+    final relay = await NostrRelay.connect(url);
+
+    // Close a connection if ALL of the following happens:
+    // 1. Has no active subscriptions
+    // 2. it has been at least 30s since last message from the relay was received
+    _initializeTimer(relay);
+    _messagesSubscription = relay.messages.listen((_) => _resetTimer());
+    _subscriptionsCountSubscription = relay.subscriptionsCountStream.listen((length) {
```
[the remark this answers is not in the document]
Comment (ice-ajax) [reply lib/app/features/nostr/providers/relays_provider.c.dart:30]: correct, fixing right now

File: lib/app/features/nostr/providers/nostr_notifier.c.dart [hunk 5]
```diff
@@ -11,14 +11,14 @@ import 'package:ion/app/features/nostr/model/nostr_entity.dart';
 import 'package:ion/app/features/nostr/providers/nostr_cache.c.dart';
 import 'package:ion/app/features/nostr/providers/nostr_event_parser.c.dart';
 import 'package:ion/app/features/nostr/providers/nostr_event_signer_provider.c.dart';
-import 'package:ion/app/features/nostr/providers/relays_provider.c.dart';
+import 'package:ion/app/features/nostr/providers/relays.c.dart';
```
[the remark this answers is not in the document]
Comment (ice-ajax) [reply lib/app/features/nostr/providers/nostr_notifier.c.dart]: fixed

**Review by ice-kreios — APPROVED** [review]

**Pull request merged**